File: five_hundred/deck.py
import random
from typing import List
from .card import Card, Suit, Rank

class Deck:
    def __init__(self):
        self.cards: List[Card] = []
        self.create_deck()

    def create_deck(self):
        """Creates a standard 43-card deck for 500."""
        self.cards = []
        all_suits = [Suit.SPADES, Suit.CLUBS, Suit.DIAMONDS, Suit.HEARTS]
        # Ranks from 5 to Ace for all suits
        value_ranks = [
            Rank.FIVE, Rank.SIX, Rank.SEVEN, Rank.EIGHT, Rank.NINE,
            Rank.TEN, Rank.JACK, Rank.QUEEN, Rank.KING, Rank.ACE
        ]
        for suit_val in all_suits:
            for rank_val in value_ranks:
                self.cards.append(Card(suit_val, rank_val)) # 40 cards

        # Add the two 4s that are kept.
        # Wikipedia: "Either the two black 4s are removed, or the 4 of spades and 4 of diamonds are removed..."
        # This means two 4s of the same color are kept (e.g. red 4s or black 4s).
        # Let's keep the 4 of Hearts and 4 of Diamonds (red 4s).
        self.cards.append(Card(Suit.HEARTS, Rank.FOUR))
        self.cards.append(Card(Suit.DIAMONDS, Rank.FOUR))
        # Add the Joker
        self.cards.append(Card(Suit.NO_TRUMP, Rank.JOKER))
        # Total: 40 (5-A) + 2 (4s) + 1 (Joker) = 43 cards.

    def shuffle(self):
        """Shuffles the cards in the deck."""
        random.shuffle(self.cards)

    def deal(self, num_cards: int) -> List[Card]:
        """Deals a specified number of cards from the top of the deck."""
        if num_cards < 0:
            raise ValueError("Number of cards to deal cannot be negative.")
        if num_cards > len(self.cards):
            raise ValueError("Not enough cards in deck to deal.")
        
        dealt_cards = self.cards[:num_cards]
        self.cards = self.cards[num_cards:]
        return dealt_cards

    def __len__(self):
        return len(self.cards)

    def __repr__(self):
        return f"Deck({len(self.cards)} cards)"
 

File: five_hundred_love/src/core/deck.lua
local Utils = require "src.core.utils"
local CardModule = require "src.core.card"
local Card = CardModule.Card
local Suit = CardModule.Suit
local Rank = CardModule.Rank

local Deck = Utils.class("Deck")

function Deck:init()
    self.cards = {}
    self:create_deck()
end

function Deck:create_deck()
    self.cards = {}
    local all_suits = {Suit.SPADES, Suit.CLUBS, Suit.DIAMONDS, Suit.HEARTS}
    local value_ranks = {
        Rank.FIVE, Rank.SIX, Rank.SEVEN, Rank.EIGHT, Rank.NINE,
        Rank.TEN, Rank.JACK, Rank.QUEEN, Rank.KING, Rank.ACE
    }
    
    -- Standard 40 cards (5-A)
    for _, suit in ipairs(all_suits) do
        for _, rank in ipairs(value_ranks) do
            table.insert(self.cards, Card.new(suit, rank))
        end
    end
    
    -- Two Red 4s (Hearts, Diamonds)
    table.insert(self.cards, Card.new(Suit.HEARTS, Rank.FOUR))
    table.insert(self.cards, Card.new(Suit.DIAMONDS, Rank.FOUR))
    
    -- Joker
    table.insert(self.cards, Card.new(Suit.NO_TRUMP, Rank.JOKER))
end

function Deck:shuffle()
    -- Standard Fisher-Yates
    for i = #self.cards, 2, -1 do
        local j = math.random(i)
        self.cards[i], self.cards[j] = self.cards[j], self.cards[i]
    end
end

function Deck:deal(num_cards)
    if num_cards < 0 then error("Cannot deal negative cards") end
    if num_cards > #self.cards then error("Not enough cards in deck") end
    
    local dealt = {}
    for i = 1, num_cards do
        -- Pop from front (or back is more efficient in Lua, but deal() usually implies top)
        -- Removing from index 1 shifts eveything, which is O(N).
        -- But N is small (43).
        table.insert(dealt, table.remove(self.cards, 1))
    end
    return dealt
end

function Deck:__len__()
    return #self.cards
end

function Deck:__tostring()
    return string.format("Deck(%d cards)", #self.cards)
end

return Deck
